### PR TITLE
Dealing with non-str keys

### PR DESCRIPTION
**Pull request by akaIDIOT**
Implementation by @bartbroere allows non-`str` keys, alternative 'fix' is to always coerce keys to `str`. One of the added tests is known to fail, we could view `_merge` as internal API and keep this function oblivious of the problem, knowing that all the `load*`-functions and the `Configuration` constructor will always call `_split_keys` which always calls the fix.

References #41, #43.

### CODE REVIEW

**Comment by akaIDIOT** [comment]
After discussion in #43, the chose solution is to explicitly reject any non-`str` key, raising errors before opening cans containing worms.

File: confidence/utils.py [hunk 1]
```diff
@@ -64,6 +68,11 @@ def _split_keys(mapping, separator='.', colliding=None):
             # recursively split key(s) in value
             value = _split_keys(value, separator)
 
+        # reject non-str keys, avoid complicating access patterns
+        if not isinstance(key, str):
```
Comment (bartbroere) [inline confidence/utils.py]: Straightforward enough :-)

File: confidence/utils.py [hunk 2]
```diff
@@ -51,6 +51,10 @@ def _split_keys(mapping, separator='.', colliding=None):
     Recursively walks *mapping* to split keys that contain the separator into
     nested mappings.
 
+    .. note::
+
+        All keys are coerced to `str` for use within a `.Configuration` object.
```
Comment (drxzcl) [inline confidence/utils.py]: Is this still correct?

Comment (akaIDIOT) [reply confidence/utils.py]: Good point, it is not. Fixed in e635d82.

File: confidence/utils.py (hunk 1 above)
Comment (akaIDIOT) [reply confidence/utils.py]: Yeah, in the end it is :P

**Pull request merged**